File: app/src/main/java/com/todobom/opennotescanner/FullScreenViewActivity.java
package com.todobom.opennotescanner;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration;
import com.nostra13.universalimageloader.core.assist.ImageSize;
import com.todobom.opennotescanner.helpers.AboutFragment;
import com.todobom.opennotescanner.helpers.Utils;
import com.todobom.opennotescanner.views.TagEditorFragment;

import java.io.File;

public class FullScreenViewActivity extends AppCompatActivity {

    private Utils utils;
    private FullScreenImageAdapter mAdapter;
    private ViewPager mViewPager;
    private AlertDialog.Builder deleteConfirmBuilder;
    private ImageLoader mImageLoader;
    private ImageSize mTargetSize;
    private int mMaxTexture;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((OpenNoteScannerApplication) getApplication()).getTracker()
                .trackScreenView("/FullScreenViewActivity", "Full Screen Viewer");

        setContentView(R.layout.activity_fullscreen_view);

        mViewPager = (ViewPager) findViewById(R.id.pager);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayShowHomeEnabled(true);
        actionBar.setTitle(null);
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_24dp);

        utils = new Utils(getApplicationContext());

        Intent i = getIntent();
        int position = i.getIntExtra("position", 0);

        // initialize Universal Image Loader
        ImageLoaderConfiguration config = new ImageLoaderConfiguration.Builder(this).build();
        mImageLoader = ImageLoader.getInstance();
        mImageLoader.init(config);

        mMaxTexture = Utils.getMaxTextureSize();
        Log.d("FullScreenViewActivity", "gl resolution: " + mMaxTexture);
        mTargetSize = new ImageSize(mMaxTexture, mMaxTexture);

        loadAdapter();

        // displaying selected image first
        mViewPager.setCurrentItem(position);

        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                Log.d("fullview", "scrolled position " + position + " offset " + positionOffset);
                Log.d("fullview", "pager " + FullScreenViewActivity.this.mViewPager.getCurrentItem());
            }

            @Override
            public void onPageSelected(int position) {
                Log.d("fullview", "selected");
                Log.d("fullview", "item" + FullScreenViewActivity.this.mViewPager.getCurrentItem());
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                Log.d("fullview", "state changed");
            }

        });

        deleteConfirmBuilder = new AlertDialog.Builder(this);

        deleteConfirmBuilder.setTitle(getString(R.string.confirm_title));
        deleteConfirmBuilder.setMessage(getString(R.string.confirm_delete_text));

        deleteConfirmBuilder.setPositiveButton(getString(R.string.answer_yes), new DialogInterface.OnClickListener() {

            public void onClick(DialogInterface dialog, int which) {
                deleteImage();
                dialog.dismiss();
            }

        });

        deleteConfirmBuilder.setNegativeButton(getString(R.string.answer_no), new DialogInterface.OnClickListener() {

            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        });

    }

    private void loadAdapter() {
        mViewPager.setAdapter(null);
        mAdapter = new FullScreenImageAdapter(FullScreenViewActivity.this,
                utils.getFilePaths());
        mAdapter.setImageLoader(mImageLoader);
        mAdapter.setMaxTexture(mMaxTexture, mTargetSize);
        mViewPager.setAdapter(mAdapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_imagepager, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        switch(id) {
            case android.R.id.home:
                finish();
                break;
            case R.id.action_tag:
                tagImage();
                return true;
            case R.id.action_share:
                shareImage();
                return true;
            case R.id.action_delete:
                deleteConfirmBuilder.create().show();
                return true;
            case R.id.action_about:
                FragmentManager fm = getSupportFragmentManager();
                AboutFragment aboutDialog = new AboutFragment();
                aboutDialog.show(fm, "about_view");
                break;
            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    private void tagImage() {
        int item = mViewPager.getCurrentItem();
        String filePath = mAdapter.getPath(item);

        FragmentManager fm = getSupportFragmentManager();
        TagEditorFragment tagEditorDialog = new TagEditorFragment();

        tagEditorDialog.setFilePath(filePath);

        tagEditorDialog.setRunOnDetach(new Runnable() {
            @Override
            public void run() {
            }
        });
        tagEditorDialog.show(fm, "tageditor_view");
    }

    private void deleteImage() {
        int item = mViewPager.getCurrentItem();

        String filePath = mAdapter.getPath(item);
        final File photoFile = new File(filePath);

        photoFile.delete();
        Utils.removeImageFromGallery(filePath,this);

        loadAdapter();
        mViewPager.setCurrentItem(item);
    }

    public void shareImage() {

        ViewPager pager = FullScreenViewActivity.this.mViewPager;
        int item = pager.getCurrentItem();

        final Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("image/jpg");
        final File photoFile = new File(mAdapter.getPath(item));
        shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(photoFile));
        Log.d("Fullscreen","uri "+Uri.fromFile(photoFile));
        startActivity(Intent.createChooser(shareIntent, getString(R.string.share_snackbar)));
    }

}

File: app/src/main/java/com/todobom/opennotescanner/FullScreenViewActivity.kt
package com.todobom.opennotescanner

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.webkit.MimeTypeMap
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.core.net.toFile
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration
import com.nostra13.universalimageloader.core.assist.ImageSize
import com.todobom.opennotescanner.helpers.AboutFragment
import com.todobom.opennotescanner.helpers.Utils
import com.todobom.opennotescanner.helpers.Utils.Companion.maxTextureSize
import com.todobom.opennotescanner.helpers.Utils.Companion.removeImageFromGallery
import com.todobom.opennotescanner.views.TagEditorFragment

class FullScreenViewActivity : AppCompatActivity() {
    private lateinit var utils: Utils
    private lateinit var mAdapter: FullScreenImageAdapter
    private lateinit var mViewPager: ViewPager
    private lateinit var deleteConfirmBuilder: AlertDialog.Builder
    private lateinit var mImageLoader: ImageLoader
    private lateinit var mTargetSize: ImageSize

    private var mMaxTexture = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_fullscreen_view)
        mViewPager = findViewById<View>(R.id.pager) as ViewPager
        val actionBar = supportActionBar!!
        actionBar.setDisplayShowHomeEnabled(true)
        actionBar.setTitle(null)
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_24dp)
        utils = Utils(applicationContext)
        val i = intent
        val position = i.getIntExtra("position", 0)

        // initialize Universal Image Loader
        val config = ImageLoaderConfiguration.Builder(this).build()
        mImageLoader = ImageLoader.getInstance()
        mImageLoader.init(config)
        mMaxTexture = maxTextureSize
        Log.d("FullScreenViewActivity", "gl resolution: $mMaxTexture")
        mTargetSize = ImageSize(mMaxTexture, mMaxTexture)
        mAdapter = loadAdapter()

        // displaying selected image first
        mViewPager.currentItem = position
        mViewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                Log.d("fullview", "scrolled position $position offset $positionOffset")
                Log.d("fullview", "pager " + mViewPager.currentItem)
            }

            override fun onPageSelected(position: Int) {
                Log.d("fullview", "selected")
                Log.d("fullview", "item" + mViewPager.currentItem)
            }

            override fun onPageScrollStateChanged(state: Int) {
                Log.d("fullview", "state changed")
            }
        })
        deleteConfirmBuilder = AlertDialog.Builder(this)
        deleteConfirmBuilder.setTitle(getString(R.string.confirm_title))
        deleteConfirmBuilder.setMessage(getString(R.string.confirm_delete_text))
        deleteConfirmBuilder.setPositiveButton(getString(R.string.answer_yes)) { dialog: DialogInterface, which: Int ->
            deleteImage()
            dialog.dismiss()
        }
        deleteConfirmBuilder.setNegativeButton(getString(R.string.answer_no)) { dialog: DialogInterface, which: Int -> dialog.dismiss() }
    }

    private fun loadAdapter(): FullScreenImageAdapter {
        mViewPager.adapter = null
        val adapter = FullScreenImageAdapter(this@FullScreenViewActivity, utils.fileUris)
        adapter.setImageLoader(mImageLoader)
        adapter.setMaxTexture(mMaxTexture, mTargetSize)
        mViewPager.adapter = adapter

        return adapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_imagepager, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        when (id) {
            android.R.id.home -> finish()
            R.id.action_tag -> {
                tagImage()
                return true
            }
            R.id.action_share -> {
                shareImage()
                return true
            }
            R.id.action_delete -> {
                deleteConfirmBuilder.create().show()
                return true
            }
            R.id.action_about -> {
                val fm = supportFragmentManager
                val aboutDialog = AboutFragment()
                aboutDialog.show(fm, "about_view")
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun isPng(context: Context, uri: Uri): Boolean {
        // First try MIME type
        val type = context.contentResolver.getType(uri)
        if (type == "image/png") return true

        // Fallback: try extension
        val extension = MimeTypeMap.getFileExtensionFromUrl(uri.toString())
        return extension.equals("png", ignoreCase = true)
    }

    private fun tagImage() {
        val item = mViewPager.currentItem
        val fileUri = mAdapter.getUri(item)
        if (isPng(this, fileUri)) {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.format_not_supported)
            builder.setMessage(R.string.format_not_supported_message)
            builder.setPositiveButton(android.R.string.ok) { dialog: DialogInterface, which: Int -> dialog.dismiss() }
            val alerta = builder.create()
            alerta.show()
            return
        }
        val tagEditorDialog = TagEditorFragment(fileUri)
        tagEditorDialog.show(supportFragmentManager, "tageditor_view")
    }

    private fun deleteImage() {
        val item = mViewPager.currentItem
        val fileUri = mAdapter.getUri(item)
        removeImageFromGallery(fileUri, this)
        loadAdapter()
        if (0 == mAdapter.count) finish()
        mViewPager.currentItem = item
    }

    fun shareImage() {
        val uri = mAdapter.getUri(mViewPager.currentItem)
        val shareIntent = Intent(Intent.ACTION_SEND)
        val shareUri = if (uri.scheme == "file") {
            FileProvider.getUriForFile(applicationContext, "$packageName.fileprovider", uri.toFile())
        } else {
            uri
        }
        shareIntent.type = this.contentResolver.getType(shareUri)
        shareIntent.putExtra(Intent.EXTRA_STREAM, shareUri)
        Log.d("Fullscreen", "uri $shareUri")
        startActivity(Intent.createChooser(shareIntent, getString(R.string.share_snackbar)))
    }
}